### PR TITLE
Features/#277 create missing env file

### PR DESCRIPTION
**Pull request by dakochik**
#277  Quick fix for missing `.yaml` / `.yml` file was implemented.

### CODE REVIEW

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateEnvFile.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    override fun getFamilyName() = SnakemakeBundle.message("INSP.NAME.conda.env.missing.fix", fileName)
+
+    override fun getText() = familyName
+
+    override fun invoke(project: Project, file: PsiFile, startElement: PsiElement, endElement: PsiElement) {
+        val tokens = fileName.split('/')
+        val targetName = tokens.lastOrNull() ?: return
+        var currentFile = file.virtualFile.parent
+        try {
+            tokens.forEach {
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateEnvFile.kt:29]: It is reasonable to expect, that in such complicated application, like IntelliJ platform the task of creating file by relative path + create intermediate dirs isn't a rare case and already solved via some API. The code below is bicycle invention in a rather complected way. E.g. java.io already have methods for creating files and folders. So here we could use smth like: 

```kotlin
        val targetFile = file.virtualFile.parent.findFileByRelativePath(fileName)
        // if targetFile == null ....
        val targetFileFolder = targetFile.parent
        VfsUtil.createDirectoryIfMissing(targetFileFolder.path)
        LocalFileSystem.getInstance().createChildFile(this, targetFileFolder, targetFile.name)
```

or just:
```kotlin
        val targetFile = Paths.get(file.virtualFile.parent.path, fileName)
        Files.createDirectories(targetFile.parent)
        Files.createFile(targetFile)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -24,6 +24,7 @@ Released on ...
 - TODO  (see [#NNN](https://github.com/JetBrains-Research/snakecharm/issues/NNN))
 
 ### Added
+- Quick fix for unresolved '.yaml' / '.yml' files (see [#277](https://github.com/JetBrains-Research/snakecharm/issues/277))
```
[the remark this answers is not in the document]
Comment (iromeo) [reply CHANGELOG.md:27]: Also could you please extend this feature not only on `conda:` section, but also for other places where we add similar file reference.

**Review by iromeo — CHANGES_REQUESTED** [review]

**Comment by iromeo** [comment]
Also it is good to make this action undoable, try to use UndoableAction interface.

**Comment by iromeo** [comment]
Allows to create files without YAML / YML / yaml / yml suffix, e.g. `envs/ddd/ddd/dddaaaa` , i think that better to forbid this and show error

**Comment by iromeo** [comment]
Error should like failed to created `/ednvs/...` file, not `untiled10`
image

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateMissedFile.kt [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        }
+        val redo = Runnable {
+            if (!supportedSections.containsKey(sectionName)) {
+                return@Runnable
+            }
+            try {
+                val directoryPath = Files.createDirectories(targetFilePath.parent)
+                val directoryVirtualFile = VfsUtil.findFile(directoryPath, true) ?: return@Runnable
+                LocalFileSystem.getInstance().createChildFile(this, directoryVirtualFile, targetFilePath.name)
+                VirtualFileManager.getInstance().asyncRefresh { }
```
[the remark this answers is not in the document]
Comment (iromeo) [reply src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateMissedFile.kt:98]: Also, not sure that we need to interact with Virtual File System here. 'undo' uses just "java.io.File" standard Java SDK API. Here maybe is better to check existence/create file w/o using VFS api, refresing, etc.

File: src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateMissedFile.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    override fun getText() = familyName
+
+    override fun invoke(
+        project: Project,
+        file: PsiFile,
+        editor: Editor?,
+        startElement: PsiElement,
+        endElement: PsiElement
+    ) {
+        val relativeDirectory =
```
Comment (iromeo) [inline src/main/kotlin/com/jetbrains/snakecharm/inspections/quickfix/CreateMissedFile.kt:64]: Fix doesn't work ok if absolute path is used, e.g.:


```
rule rule277_s:
    conda: "/ednvs/ddd/ddd/dddaaaa.yaml"
    script: "env/ddddd.py"
```

on runtime expects absolute path:
```
$ snakemake --snakefile rule277.smk --use-conda -c1
Building DAG of jobs...
WorkflowError:
Failed to open source file /ednvs/ddd/ddd/dddaaaa.yaml
FileNotFoundError: [Errno 2] No such file or directory: '/ednvs/ddd/ddd/dddaaaa.yaml'
```

but quick fixes tries to create a file in project working directory. I assume that it valid for all file references for conda, script, etc sections. E.g. path could be absolute or relative. Is better to check.

**Review by iromeo — COMMENTED** [review]
Let's disscuss some impl details + need to fix absolute paths issue. Optionally you could file a new issue for abs path support and do not include it into this PR.

**Comment by dakochik** [comment]
Discussed problems were solved.

**Pull request merged**